From 54434c6cda3169f518279cb8264dc08c144de586 Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:39:21 -0500
Subject: [drm] i915: Handle failure to kick out a conflicting fb driver

Message-id: <1418852380-13061-71-git-send-email-rclark@redhat.com>
Patchwork-id: 102228
O-Subject: [RHEL7 drm 70/89] drm/i915: Handle failure to kick out a conflicting fb driver
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: since drm-intel-next-2014-07-25

commit f96de58fc7e7d3d717c7c63975c3b896c906b5e3

Author:     Chris Wilson <chris@chris-wilson.co.uk>
AuthorDate: Mon Dec 16 15:57:40 2013 +0000
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Mon Dec 15 14:18:35 2014 -0500

    drm/i915: Handle failure to kick out a conflicting fb driver

    Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
    Reviewed-by: Jani Nikula <jani.nikula@intel.com>
    Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/i915_dma.c b/drivers/gpu/drm/i915/i915_dma.c
index be291d0..644a7a9 100644
--- a/drivers/gpu/drm/i915/i915_dma.c
+++ b/drivers/gpu/drm/i915/i915_dma.c
@@ -1425,15 +1425,16 @@ void i915_master_destroy(struct drm_device *dev, struct drm_master *master)
 }
 
 #if IS_ENABLED(CONFIG_FB)
-static void i915_kick_out_firmware_fb(struct drm_i915_private *dev_priv)
+static int i915_kick_out_firmware_fb(struct drm_i915_private *dev_priv)
 {
  struct apertures_struct *ap;
  struct pci_dev *pdev = dev_priv->dev->pdev;
  bool primary;
+ int ret;
 
  ap = alloc_apertures(1);
  if (!ap)
-  return;
+  return -ENOMEM;
 
  ap->ranges[0].base = dev_priv->gtt.mappable_base;
  ap->ranges[0].size = dev_priv->gtt.mappable_end;
@@ -1441,13 +1442,16 @@ static void i915_kick_out_firmware_fb(struct drm_i915_private *dev_priv)
  primary =
   pdev->resource[PCI_ROM_RESOURCE].flags & IORESOURCE_ROM_SHADOW;
 
- remove_conflicting_framebuffers(ap, "inteldrmfb", primary);
+ ret = remove_conflicting_framebuffers(ap, "inteldrmfb", primary);
 
  kfree(ap);
+
+ return ret;
 }
 #else
-static void i915_kick_out_firmware_fb(struct drm_i915_private *dev_priv)
+static int i915_kick_out_firmware_fb(struct drm_i915_private *dev_priv)
 {
+ return 0;
 }
 #endif
 
@@ -1664,7 +1668,11 @@ int i915_driver_load(struct drm_device *dev, unsigned long flags)
    goto out_gtt;
   }
 
-  i915_kick_out_firmware_fb(dev_priv);
+  ret = i915_kick_out_firmware_fb(dev_priv);
+  if (ret) {
+   DRM_ERROR("failed to remove conflicting framebuffer drivers\n");
+   goto out_gtt;
+  }
  }
 
  pci_set_master(dev->pdev);
-- 
1.7.1